esrally/actor.py

```diff
@@ -85,16 +85,27 @@ def receiveMsg_DefuseBomb(self, msg, sender):
     :param actor_name: A human readable name of the current actor that should be used in the exception message.
     """
 
+    def _format_exception(e):
+        clazz = type(e)
+        message = getattr(e, "message", lambda: e)
+        return f"{'.'.join([clazz.__module__, clazz.__name__])}: {message}"
+
     def guard(self, msg, sender):
         try:
             return f(self, msg, sender)
         except BaseException as e:
-            msg = "Error in {}".format(actor_name)
             # log here as the full trace might get lost.
-            logging.getLogger(__name__).exception(msg)
+            logging.getLogger(__name__).exception("Error in %s", actor_name)
             # don't forward the exception as is because the main process might not have this class available on the load path
             # and will fail then while deserializing the cause.
-            self.send(sender, BenchmarkFailure("{} ({})".format(msg, str(e))))
+            parsed_exception = _format_exception(e)
+            nesting = 0
+            while hasattr(e, "cause") and e.cause:
+                nesting += 1
+                e = e.cause
+                parsed_exception += f" (Caused by: {_format_exception(e)}"
+            parsed_exception += nesting * ")"
```

Comment:
What are your thoughts on using the traceback module?

```suggestion
            parsed_exception = traceback.format_exc()
```

It's shorter and simpler. It does print a traceback to stdout, but that's to be expected when you run a Python program from the command line? (Not asking you to change this, just suggesting the idea. I know you didn't want to print everything to stdout initially.)

Reply:
results in 
```
 % esrally race --track-path ../rally-tracks/http_logs --test-mode

    ____        ____
   / __ \____ _/ / /_  __
  / /_/ / __ `/ / / / / /
 / _, _/ /_/ / / / /_/ /
/_/ |_|\__,_/_/_/\__, /
                /____/

[INFO] Race id is [c1f21a58-d307-4493-8c37-487d69b6c183]
[ERROR] Cannot race. Traceback (most recent call last):
  File "/Users/rick.boyd/dev/rally/esrally/track/loader.py", line 982, in read
    rendered = render_template_from_file(track_spec_file, self.track_params, complete_track_params=self.complete_track_params)
  File "/Users/rick.boyd/dev/rally/esrally/track/loader.py", line 781, in render_template_from_file
    return render_template(
  File "/Users/rick.boyd/dev/rally/esrally/track/loader.py", line 750, in render_template
    return template.render()
  File "/Users/rick.boyd/dev/rally/.venv/lib/python3.8/site-packages/jinja2/environment.py", line 1090, in render
    self.environment.handle_exception()
  File "/Users/rick.boyd/dev/rally/.venv/lib/python3.8/site-packages/jinja2/environment.py", line 832, in handle_exception
    reraise(*rewrite_traceback_stack(source=source))
  File "/Users/rick.boyd/dev/rally/.venv/lib/python3.8/site-packages/jinja2/_compat.py", line 28, in reraise
    raise value.with_traceback(tb)
  File "<template>", line 48, in top-level template code
jinja2.exceptions.UndefinedError: 'runtime' is undefined

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/rick.boyd/dev/rally/esrally/actor.py", line 96, in guard
    return f(self, msg, sender)
  File "/Users/rick.boyd/dev/rally/esrally/racecontrol.py", line 110, in receiveMsg_Setup
    self.coordinator.setup(sources=msg.sources)
  File "/Users/rick.boyd/dev/rally/esrally/racecontrol.py", line 194, in setup
    self.current_track = track.load_track(self.cfg)
  File "/Users/rick.boyd/dev/rally/esrally/track/loader.py", line 192, in load_track
    return _load_single_track(cfg, repo, repo.track_name)
  File "/Users/rick.boyd/dev/rally/esrally/track/loader.py", line 199, in _load_single_track
    current_track = reader.read(track_name, track_repository.track_file(track_name), track_dir)
  File "/Users/rick.boyd/dev/rally/esrally/track/loader.py", line 1007, in read
    raise TrackSyntaxError(msg, e)
esrally.track.loader.TrackSyntaxError: Could not load '/Users/rick.boyd/dev/rally-tracks/http_logs/track.json'. The complete track has been written to '/var/folders/_j/86s8k_jd4md2k6ch_290_0mr0000gn/T/tmpyjca3iqo.json' for diagnosis.


Getting further help:
*********************
* Check the log files in /Users/rick.boyd/.rally/logs for errors.
* Read the documentation at https://esrally.readthedocs.io/en/latest/.
* Ask a question on the forum at https://discuss.elastic.co/tags/c/elastic-stack/elasticsearch/rally.
* Raise an issue at https://github.com/elastic/rally/issues and include the log files in /Users/rick.boyd/.rally/logs.

--------------------------------
[INFO] FAILURE (took 30 seconds)
--------------------------------
```
What do you (and @danielmitterdorfer) think?

Reply:
yeah... i'm good with it

Reply:
(PEP 654 was indeed accepted for Python 3.11: https://discuss.python.org/t/accepting-pep-654-exception-groups-and-except/10813)